### PR TITLE
Adding developer documentation for ParallelAccelerator technology.

### PR DESCRIPTION
**Pull request by DrTodd13**

### CODE REVIEW

**Comment by seibert** [comment]
Looks like there is a ReST error with an underline not being long enough under a heading.  (ReST cares about this, apparently.)

File: docs/source/developer/architecture.rst [hunk 1]
```diff
@@ -365,7 +365,7 @@ types, language features, or functions are used in the function body.
 
 .. _`rewrite-typed-ir`:
 
-Stage 6: Rewrite typed IR
+Stage 6a: Rewrite typed IR
 -------------------------
```
Comment (seibert) [inline docs/source/developer/architecture.rst]: Need another dash on this line

File: docs/source/developer/architecture.rst [hunk 2]
```diff
@@ -431,14 +431,177 @@ ufunc-like function that is inlined into a single loop that only
...
+are described in more detail in the following paragraphs.
+
+The automatic parallelization pass has a number of sub-passes.
+
+1. CFG simplification - Sometimes Numba IR will contain chains of blocks
+containing no loops which are merged in this sub-pass into single blocks.
+This sub-pass simplifies subsequent analysis of the IR.
+
+2. Numpy canonicalization - Some Numpy operations can be written as
+operations on Numpy objects (e.g., arr.sum()), or as calls to Numpy
```
Comment (DrTodd13) [inline docs/source/developer/architecture.rst]: Will fix.

File: docs/source/developer/architecture.rst [hunk 3]
```diff
@@ -431,14 +431,177 @@ ufunc-like function that is inlined into a single loop that only
...
+the number of parfors that are adjacent to each other such that they can
+then be fused together into one parfor that takes only one pass over the
+data and will thus typically have better cache performance.  Finally,
+during lowering, these parfor operators are converted to a form similar
+to guvectorize to implement the actual parallelism.  All these processes
+are described in more detail in the following paragraphs.
+
+The automatic parallelization pass has a number of sub-passes.
+
+1. CFG simplification - Sometimes Numba IR will contain chains of blocks
```
Comment (DrTodd13) [inline docs/source/developer/architecture.rst]: Done.

File: docs/source/developer/architecture.rst [hunk 4]
```diff
@@ -431,14 +431,177 @@ ufunc-like function that is inlined into a single loop that only
...
+containing no loops which are merged in this sub-pass into single blocks.
+This sub-pass simplifies subsequent analysis of the IR.
+
+2. Numpy canonicalization - Some Numpy operations can be written as
+operations on Numpy objects (e.g., arr.sum()), or as calls to Numpy
+taking those objects (e.g., numpy.sum(arr)).  This sub-pass converts all
```
Comment (DrTodd13) [inline docs/source/developer/architecture.rst]: Done.

File: docs/source/developer/architecture.rst [hunk 5]
```diff
@@ -431,14 +431,177 @@ ufunc-like function that is inlined into a single loop that only
...
+such operations to the latter form for cleaner subsequent analysis.
+
+3. Array analysis - A critical requirement for later parfor fusion is that 
+parfors have identical iteration spaces and these iteration spaces
```
Comment (DrTodd13) [inline docs/source/developer/architecture.rst]: Actually, this is a benefit of our parfor construction that we do not require what might be called classic loop dependence analysis.  Array analysis is a very different beast than loop dependence analysis.

File: docs/source/developer/architecture.rst (hunk 3 above)
Comment (DrTodd13) [inline docs/source/developer/architecture.rst]: Without blank lines between, it puts it all in one paragraph for me.  Indent didn't help either.  Please help here.

File: docs/source/developer/architecture.rst [hunk 6]
```diff
@@ -431,14 +431,177 @@ ufunc-like function that is inlined into a single loop that only
...
+typically correspond to the sizes of the dimensions of Numpy arrays.
+In this sub-pass, the IR is analyzed to determine equivalence classes for
+the dimensions of Numpy arrays.  Consider the example, a = b + 1, where `a`
```
Comment (DrTodd13) [inline docs/source/developer/architecture.rst]: Done.

File: docs/source/developer/architecture.rst [hunk 7]
```diff
@@ -431,14 +431,177 @@ ufunc-like function that is inlined into a single loop that only
...
+and `b` are both Numpy arrays.  Here, we know that each dimension of `a` 
+must have the same equivalence class as the corresponding dimension of `b`.
+Typically, routines rich in Numpy operations will enable equivalence classes
+to be fully known for all arrays created within a function.
+
+4. prange to parfor - The use of prange in a for loop is an explicit
```
Comment (DrTodd13) [inline docs/source/developer/architecture.rst]: Done.

File: docs/source/developer/architecture.rst [hunk 8]
```diff
@@ -431,14 +431,177 @@ ufunc-like function that is inlined into a single loop that only
...
+indication from the programmer that all iterations of the for loop can
+execute in parallel.  In this sub-pass, we analyze the CFG to locate loops
+and to convert those loops controlled by a prange object to the explicit 
+`parfor` operator.  Each explicit parfor operator consists of:
+
+a. A list of loop nest information that describes the iteration space of the parfor.
+Each entry in the loop nest list contains an indexing variable, the start of
+the range, the end of the range, and the step value for each iteration.
+
+b. An init block which contains instructions to be executed one time before
```
Comment (DrTodd13) [inline docs/source/developer/architecture.rst]: Done.

File: docs/source/developer/architecture.rst [hunk 9]
```diff
@@ -431,14 +431,177 @@ ufunc-like function that is inlined into a single loop that only
...
+`parfor` operator.  Each explicit parfor operator consists of:
+
+a. A list of loop nest information that describes the iteration space of the parfor.
+Each entry in the loop nest list contains an indexing variable, the start of
+the range, the end of the range, and the step value for each iteration.
+
+b. An init block which contains instructions to be executed one time before 
+the parfor begins executing.
+
+c. The loop body is a set of basic blocks that correspond to the body of the
```
Comment (DrTodd13) [inline docs/source/developer/architecture.rst]: Done.

File: docs/source/developer/architecture.rst [hunk 10]
```diff
@@ -431,14 +431,177 @@ ufunc-like function that is inlined into a single loop that only
...
+loop and compute one point in the iteration space.
+
+d. The index variables used for each dimension of the iteration space.
+
+For parfor pranges, the loop nest is a single entry where the start, stop, and
+step fields come from the specified prange.  The init block is empty for 
+prange parfors and the loop body is the set of blocks in the loop minus the
+loop header.
+
+5. Numpy to parfor - In this sub-pass, Numpy functions such as `ones`,
```
Comment (DrTodd13) [inline docs/source/developer/architecture.rst]: Done.

File: docs/source/developer/architecture.rst [hunk 11]
```diff
@@ -431,14 +431,177 @@ ufunc-like function that is inlined into a single loop that only
...
+the number of dimensions of the left-hand side of the assignment instruction in
+the IR.  The number and size of the dimensions of the left-hand-side array is
+taken from the array analysis information generated in sub-pass 3 above.
+An instruction to create the result Numpy array is generated and stored in the
+new parfor's init block.  A basic block is created for the loop body and an
+instruction is generated and added to the end of that block to store the result
+of the computation into the array at the current point in the iteration space.
+The result stored into the array depends on the operation that is being 
+converted.  For example, for `ones`, the value stored is a constant 1.  For
+calls to generate random array, the value comes from a call to the same
```
Comment (DrTodd13) [inline docs/source/developer/architecture.rst]: Done.

File: docs/source/developer/architecture.rst [hunk 12]
```diff
@@ -493,6 +656,97 @@ generated above quite significantly:
...
+this stage creates a Python function as text that contains the parameters
+to the GUFunc and iteration code that takes the current schedule entry
+and loops through the specified portion of the iteration space.  In the
+body of that loop, a special sentinel is inserted for subsequent easy
+location.  This code that handles the processing of the iteration space
+is then eval'ed into existence and the Numba compiler's run_frontend
+function is called to generate IR.  That IR is scanned to locate the
+sentinel and the sentinel is replaced with the loop body of the parfor.
+Then, the process of creating the parallel GUFunc is completed by
+compiling this merged IR with the Numba compiler's compile_ir function.
```
Comment (DrTodd13) [inline docs/source/developer/architecture.rst]: Done.

File: docs/source/developer/architecture.rst [hunk 13]
```diff
@@ -493,6 +656,97 @@ generated above quite significantly:
...
+
+To call the parallel GUFunc, the static schedule must be created.
+Code is inserted to call a function named `do_scheduling.`  This function
```
Comment (DrTodd13) [inline docs/source/developer/architecture.rst]: Done.

File: docs/source/developer/architecture.rst [hunk 14]
```diff
@@ -493,6 +656,97 @@ generated above quite significantly:
...
+is called with the size of each of the parfor's dimensions and the number
+`N` of configured Numba threads.  The do_scheduling function will divide
```
Comment (DrTodd13) [inline docs/source/developer/architecture.rst]: Done.

File: docs/source/developer/architecture.rst [hunk 15]
```diff
@@ -493,6 +656,97 @@ generated above quite significantly:
...
+dedicated to a given dimension of the full iteration space is roughly
+proportional to the ratio of the size of the given dimension to the sum
+of the sizes of all the dimensions of the iteration space.
+
+Parallel reductions are not natively provided by GUFuncs but the parfor
+lowering strategy allows us to use GUFuncs in a way that reductions can
+be performed in parallel.  To accomplish this, for each reduction variable
+computed by a parfor, the parallel GUFunc and the code that calls it is
+modified to make the scalar reduction variable into an array of reduction
+variable whose length is equal to the number of Numba threads.  In addition,
```
Comment (DrTodd13) [inline docs/source/developer/architecture.rst]: Done.

File: docs/source/reference/envvars.rst [hunk 16]
```diff
@@ -88,6 +88,11 @@ These variables influence what is printed out during compilation of
    its CPython-compatible wrapper (whose name begins with ``wrapper.``).
    Note that the function is often inlined inside the wrapper, as well.
 
+.. envvar:: NUMBA_DEBUG_ARRAY_OPT
+
+   Dump debugging information related to the processing associated with
+   the parallel=True jit decorator option.
```
Comment (DrTodd13) [inline docs/source/reference/envvars.rst]: Done.

File: docs/source/developer/architecture.rst [hunk 17]
```diff
@@ -431,14 +431,177 @@ ufunc-like function that is inlined into a single loop that only
...
+
+b. An init block which contains instructions to be executed one time before 
+the parfor begins executing.
+
+c. The loop body is a set of basic blocks that correspond to the body of the
+loop and compute one point in the iteration space.
+
+d. The index variables used for each dimension of the iteration space.
+
+For parfor pranges, the loop nest is a single entry where the start, stop, and
```
Comment (DrTodd13) [inline docs/source/developer/architecture.rst]: Done.

File: docs/source/developer/architecture.rst [hunk 18]
```diff
@@ -431,14 +431,177 @@ ufunc-like function that is inlined into a single loop that only
...
+instruction is generated and added to the end of that block to store the result
+of the computation into the array at the current point in the iteration space.
+The result stored into the array depends on the operation that is being 
+converted.  For example, for `ones`, the value stored is a constant 1.  For
+calls to generate random array, the value comes from a call to the same
+random number function but with the size parameter dropped and therefore
+returning a scalar.  For arrayexpr operators, the arrayexpr tree is converted
+to Numba IR and the value at the root of that expression tree is used to
+write into the output array.
+For reductions, the loop nest list is similarly created using the array
```
Comment (DrTodd13) [inline docs/source/developer/architecture.rst]: Done.

File: docs/source/developer/architecture.rst [hunk 19]
```diff
@@ -431,14 +431,177 @@ ufunc-like function that is inlined into a single loop that only
...
+the second parfor.  The two parfors are fused together by adding the 
+second parfor's init block to the first's, merging the two parfors'
+loop bodies together and replacing the instances of the second parfor's
+loop index variables in the second parfor's body with the loop index
+variables for the first parfor.
+
+8. Push call objects and compute parfor parameters - 
+In the lowering phase described in Section
+:ref:`lowering`, each parfor becomes a separate function executed in
+parallel in the guvectorize style.  Since parfors may use variables
```
Comment (DrTodd13) [inline docs/source/developer/architecture.rst]: Done.

File: docs/source/developer/architecture.rst (hunk 5 above)
Comment (stuartarchibald) [inline docs/source/developer/architecture.rst]: Thanks, I obviously didn't make the differentiation. Is there a seminal reference for this array analysis method that could perhaps be added?

**Comment by stuartarchibald** [comment]
Thanks for the fixups @DrTodd13 . I've formatted the large list section as a list which abuses `enumerated list` with `definition list` syntax to achieve a numbered list with definition headers for each step and then alpha-enumerated sublist, patch is here, https://github.com/stuartarchibald/numba/commit/51169b51732679077399bb4fe969bc5f8fb16e98. If you prefer it feel free to pull it in.

**Pull request merged**